grpc/runtime/src/main/scala/io/buoyant/grpc/runtime/DecodingStream.scala

```diff
@@ -0,0 +1,382 @@
+
+    case class Buffer(
+      /** The current gRPC message header, if one has been parsed */
+      header: Option[Header] = None,
+      /** Unparsed bytes */
+      buf: Buf = Buf.Empty,
+      /** Tracks how many bytes are consumed and when the underling data may be released */
+      releaser: Releaser = Releaser.Nil
+    ) extends RecvState
+
+    val Empty: RecvState = Buffer()
+
+    // TODO Introduce an additional state to indicate when the stream
+    // has been reset.
+  }
+
+  /**
+   * Releasers track releasable segments across regions of an
+   * h2.Stream.
+   *
+   * Frames added to `track` are appended to byte tracking.  As
+   * headers & messages are parsed, `consume` is called to mark bytes
+   * as pending to be released.  Finally, `releasable` returns a
+   * function that can be used to release pending data (since the last
+   * call to releasable).
+   */
+  private trait Releaser {
+
+    /**
+     * Add the given H2 frame to be tracked and released.
+     */
+    def track(frame: h2.Frame): Releaser
+
+    /**
+     * Mark `n` bytes as consumed.
+     */
+    def consume(n: Int): Releaser
+
+    /**
+     * Reap all consumed frames into a releasable function, and return
+     * the updated state afterwards.
+     */
+    def releasable(): (Releaser, Releaser.Func)
+  }
+
+  private object Releaser {
+    type Func = () => Future[Unit]
+    val NopFunc: Func = () => Future.Unit
+
+    object Underflow
+      extends Exception("released too many bytes")
+      with NoStackTrace
+
+    object Nil extends Releaser {
+      override def toString = "Releaser.Nil"
+      override def track(f: h2.Frame): Releaser = mk(f)
+
+      override def consume(n: Int): Releaser =
+        if (n == 0) this
+        else throw Underflow
+
+      override def releasable(): (Releaser, Func) = _releasable
+      private[this] val _releasable = (this, NopFunc)
+    }
+
+    private[this] object FrameReleaser {
+
+      /**
+       * Tracks how many segments of a single frame need to be
+       * released and, when all have been released, calls `underlying`
+       * to release the frame.
+       */
+      trait Latch {
+        def segmentReleaseFunc(): Func
+        def lastSegmentReleaseFunc(): Func
+      }
+
+      object Latch {
+        private[this] trait State
+        private[this] case class Incomplete(pending: Int) extends State {
+          require(pending >= 0)
+          def decrement = copy(pending = pending - 1)
+          def increment = copy(pending = pending + 1)
+        }
+        private[this] case class Releasable(pending: Int) extends State {
+          require(pending >= 0)
+          def decrement = copy(pending = pending - 1)
+        }
+
+        private[this] class Segmented(underlying: Func) extends Latch {
+          @volatile private[this] var canReleaseUnderlying = false
+          @volatile private[this] var pending = 0
+
+          private[this] val releaseSegment: Func = { () =>
```

Comment:
If I'm following this correctly, I think this can be broken into it's two codepaths and put directly into `segmentReleaseFunc` and `lastSegmentReleaseFunc`.  I think the `canReleaseUnderyling` variable is unnecessary